File: day1/main.go
package main

import (
	"bytes"
	"fmt"
	"log"
	"os"
	"sort"
	"strconv"
)

func main() {

	// Make a structure to keep track of elf totals
	type Elf struct {
		number int
		cals   int
	}
	elves := make([]Elf, 0)

	// Read the data into the array, converting from strings to integers
	data, err := os.ReadFile("input.dat")
	if err != nil {
		log.Fatal(err)
	}

	// Split the data on blank lines
	calorieData := bytes.Split(data, []byte{'\n', '\n'})
	for i, token := range calorieData {

		// New elf
		elf := Elf{number: i + 1, cals: 0}

		// Calculate all this elf's calories
		for _, entry := range bytes.Split(token, []byte{'\n'}) {
			cals, err := strconv.Atoi(string(entry))
			if err != nil {
				log.Fatal(err)
			}
			elf.cals += cals
		}

		// Add this elf to the array
		elves = append(elves, elf)
	}

	// Sort the elf array descending by total calories
	sort.Slice(elves, func(i, j int) bool {
		return elves[i].cals > elves[j].cals
	})

	// Part 1
	fmt.Printf("Part1: Elf %d has %d calories\n", elves[0].number, elves[0].cals)

	// Part2
	total := 0
	for i := 0; i < 3; i++ {
		total += elves[i].cals
	}
	fmt.Printf("Part2: Top three elves have %d calories\n", total)
}

File: day1/Main.java
import java.io.File;
import java.io.IOException;
import java.util.List;
import java.util.ArrayList;
import java.util.Scanner;

// Define a data structure for an Elf.
class Elf {
    int number;
    int cals;

    public Elf(int number, int cals) {
        this.number = number;
        this.cals = cals;
    }
}

public class Main {
    public static void main(String[] args) throws IOException {
        // Make a list to keep track of elf numbers and their total calories.
        List<Elf> elves = new ArrayList<Elf>();

        // Read the input data into a list, with a blank line separating the entries
        // between the calorie records for each elf.
        Scanner scanner = new Scanner(new File("input.dat"));
        scanner.useDelimiter("\n\n");
        while (scanner.hasNext()) {
            String entries = scanner.next();
            int i = elves.size() + 1;
            Elf elf = new Elf(i, 0);
            // Calculate all this elf's calories.
            Scanner entryScanner = new Scanner(entries);
            while (entryScanner.hasNext()) {
                int cals = entryScanner.nextInt();
                elf.cals += cals;
            }
            entryScanner.close();
            // Add this elf to the list.
            elves.add(elf);
        }
        scanner.close();

        // Sort the elf list descending by total calories.
        elves.sort((e1, e2) -> Integer.compare(e2.cals, e1.cals));

        // Part 1
        Elf topElf = elves.get(0);
        System.out.printf("Part1: Elf %d has %d calories\n", topElf.number, topElf.cals);

        // Part 2
        int total = elves.get(0).cals + elves.get(1).cals + elves.get(2).cals;
        System.out.printf("Part2: Top three elves have %d calories\n", total);
    }
}

